api/src/main/java/edu/wpi/first/shuffleboard/api/data/DataTypes.java

```diff
@@ -1,95 +1,225 @@
+  public static void setDefault(DataTypes instance) {
+    TestUtils.assertRunningFromTest();
+    defaultInstance = instance;
   }
 
   /**
    * Registers the given data type.
    *
    * @param dataType the data type to register
    */
-  public static void register(DataType<?> dataType) {
+  @Override
+  public void register(DataType dataType) {
+    requireNonNull(dataType, "dataType");
+    if (isRegistered(dataType)) {
+      throw new IllegalArgumentException("Data type " + dataType + " has already been registered");
+    }
     dataTypes.put(dataType.getName(), dataType);
+    typeCache.put(dataType.getJavaClass(), Optional.of(dataType));
+    addItem(dataType);
+  }
+
+  @Override
+  public void unregister(DataType dataType) {
+    dataTypes.remove(dataType.getName());
+    typeCache.remove(dataType.getJavaClass());
+    removeItem(dataType);
   }
 
   /**
    * Gets the data type with the given name.
    */
-  public static Optional<DataType<?>> forName(String name) {
+  public Optional<DataType> forName(String name) {
     return Optional.ofNullable(dataTypes.get(name));
   }
 
   /**
+   * Gets the data type most relevant to a Java class.
+   */
+  @SuppressWarnings("unchecked")
+  public <T> Optional<DataType<T>> forJavaType(Class<T> type) {
+    if (type.isPrimitive()) {
+      return forJavaType((Class) Primitives.wrap(type));
+    }
+    return (Optional) typeCache.computeIfAbsent(type, __ -> {
+      if (DataType.class.isAssignableFrom(type)) {
+        return forType((Class<DataType>) type);
+      }
+      Optional<DataType> naive = dataTypes.values().stream()
+          .filter(t -> type.equals(t.getJavaClass()))
+          .findAny();
+      if (naive.isPresent()) {
+        return naive;
+      } else {
+        // Check Java class hierarchy
+        Comparator<Class<?>> classComparator = closestTo(type);
+        List<DataType> sorted = dataTypes.values().stream()
+            .filter(t -> t != All)
+            .sorted((t1, t2) -> classComparator.reversed().compare(t1.getJavaClass(), t2.getJavaClass()))
+            .collect(Collectors.toList());
+        return Optional.of(sorted.get(0));
+      }
+    });
+  }
+
+  /**
+   * Creates a comparator object that compares classes in order of closest to the target class, in terms of class
+   * hierarchy. For example, take a class hierarchy of
+   * <pre><code>
+   *      Foo
+   *     /   \
+   *   Bar   Baz
+   *    |     |
+   *  Object Buq
+   *          |
+   *        Object
+   * </code></pre>
+   *
+   * {@code closestTo(Object.class)} would be sorted as {@code [Bar, Buq], Baz, Foo}. The order of Bar, Baz is not
+   * deterministic and depends on the ordering of the source collection, hence the brackets.
+   *
+   * <p><b>This method does <i>not</i> support comparison of interfaces</b></p>
+   *
+   * @param target the class that should be compared to to determine ordering
+   */
+  @VisibleForTesting
+  static Comparator<Class<?>> closestTo(Class<?> target) {
```

Comment:
I disagree with that mentality. It makes sense to write tests where you need to make sure something works correctly.